### PR TITLE
Tests, Docs, Decode/Encode Special Chars

### PR DESCRIPTION
**Pull request by charliesantos**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

## Pull Request Details
- unit and integration tests
- encode/decode special characters
- doc updates

### Before review
* [x] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [x] Updated affected documentation
* [x] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

File: tests/integration/connectToken.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  function waitFor(n: number, reject?: boolean) {
+    return new Promise((res, rej) => setTimeout(reject ? rej : res, n));
+  }
+
+  function toObject(map: Map<string, string>) {
+    return Array.from(map.keys()).reduce((result: Record<string, string>, key: string) => {
+      result[key] = map.get(key)!;
+      return result;
+    }, {})
```
Comment (mhuynh5757) [inline tests/integration/connectToken.ts:19]: You may be able to solve this by using something simpler
```
return Object.fromEntries(map.entries());
```

File: tests/integration/connectToken.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  let callerDeviceA: Device;
+  let callerDeviceB: Device;
+  let receiverDevice: Device;
+  let reconnecDeviceA: Device;
```
Comment (mhuynh5757) [inline tests/integration/connectToken.ts:25]:
```suggestion
  let reconnectDeviceA: Device;
```
Reconnect is misspelled in several places in this file.

File: tests/unit/call.ts [hunk 3]
```diff
@@ -107,6 +107,42 @@ describe('Call', function() {
       assert.equal(conn.customParameters.get('baz'), 123);
     });
 
+    context('connectToken', () => {
+      it('should return underfined if callsid is missing', () => {
```
Comment (mhuynh5757) [inline tests/unit/call.ts:111]:
```suggestion
      it('should return undefined if callsid is missing', () => {
```

File: tests/unit/call.ts [hunk 4]
```diff
@@ -107,6 +107,42 @@ describe('Call', function() {
...
+        conn = new Call(config, Object.assign(options, { reconnectToken: 'testReconnectToken' }));
+        assert.equal(conn.connectToken, undefined);
+      });
+
+      it('should return underfined if reconnectToken is missing', () => {
```
Comment (mhuynh5757) [inline tests/unit/call.ts:116]:
```suggestion
      it('should return undefined if reconnectToken is missing', () => {
```

File: tests/unit/device.ts [hunk 5]
```diff
@@ -176,29 +177,100 @@ describe('Device', function() {
...
+            sinon.assert.calledWithExactly(publisher.setHost, 'eventgw.twilio.com');
+          });
 
-        it('should set default host address if home is not available', () => {
-          pstream.emit('connected', {});
-          sinon.assert.calledOnce(publisher.setHost);
-          sinon.assert.calledWithExactly(publisher.setHost, 'eventgw.twilio.com');
+          Object.values(Region).forEach(region => {
+            it(`should set host to eventgw.${region}.twilio.com when home is set to ${region}`, () => {
+              pstream.emit('connected', { home: region});
```
Comment (mhuynh5757) [inline tests/unit/device.ts:200]:
```suggestion
              pstream.emit('connected', { home: region });
```

**Review by mhuynh5757 — CHANGES_REQUESTED** [review]

File: tests/integration/connectToken.ts (hunk 1 above)
Comment (charliesantos) [reply tests/integration/connectToken.ts:19]: I thought I was getting compilation error when I tried this. I'll take a look again.

Comment (mhuynh5757) [reply tests/integration/connectToken.ts:19]: Yeah no big deal. Let's shelf it.

**Review by mhuynh5757 — APPROVED** [review]

**Pull request merged**